### PR TITLE
Parse TOML timestamps as strings

### PR DESCRIPTION
**Pull request by andreabedini**
# Motivation

Writing [foliage](https://github.com/andreabedini/foliage/) I have choosen TOML knowing that nix would be able to read it natively. As it turns out, I cannot parse foliage input metadata because it includes timestamps, which nix does not support. :broken_heart: 

# Context

Currently fromTOML throws an exception when encountering a timestamp since the nix language lacks a way to represent them. This patch changes this beaviour and makes fromTOML parse timestamps as their were strings.

I believe this is the right thing to do here, given that a string is better than nothing.

Note that, ifwe had `toTOML`, ensuring `fromTOML (toTOML x) == x` might be impossible anyway because nix values can be expressed in multiple ways in TOML.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

### CODE REVIEW

**Comment by Ericson2314** [comment]
Discussed at last Nix team meeting.

Verdict: Let's do this, but in a way that round-trips properly, and as an experimental feature.

<details>
- @edolstra: Probably the best we can do. The alternative would be encoding it in a structured way, which would be fairly complicated.

- @roberth: Parsing it in a structured way might actually be cool. In particular for NixOS use-case.
  Middle-ground: parse it as `{ _type = "timestamp"; value = "${theRawTSString}"; }`

- @Ericson2314: Should probably make it experimental, then we can refine it over time, and unblock the person without stabilizing anything rashly.

- @roberth: Sure, but I don't mind stabilizing it sooner.
</details>

**Comment by andreabedini** [comment]
:+1:  I think we should follow the TOML's definitions ([Offset Date-Time](https://toml.io/en/v1.0.0#offset-date-time), [Local Date-Time](https://toml.io/en/v1.0.0#local-date-time), [Local Date](https://toml.io/en/v1.0.0#local-date), [Local Date](https://toml.io/en/v1.0.0#local-date)). These types are already available from toml++. AFAIK they are all subsets of RFC 3339 anyway.

With some luck I'll find some time to work on this week or the next.

File: src/libexpr/primops/fromTOML.cc [hunk 1]
```diff
@@ -57,10 +59,15 @@ static void prim_fromTOML(EvalState & state, const PosIdx pos, Value * * args, V
             case toml::value_t::local_datetime:
             case toml::value_t::offset_datetime:
             case toml::value_t::local_date:
-            case toml::value_t::local_time:
-                // We fail since Nix doesn't have date and time types
-                throw std::runtime_error("Dates and times are not supported");
+            case toml::value_t::local_time: {
+                auto attrs = state.buildBindings(2);
+                attrs.alloc("_type").mkString("timestamp");
```
[the remark this answers is not in the document]
Comment (fricklerhandwerk) [reply src/libexpr/primops/fromTOML.cc:64]: Related: https://github.com/NixOS/nix/issues/3929

**Comment by fricklerhandwerk** [comment]
Reviewed in Nix team meeting 2023-05-08:

We still want this, but we also want to get it right. As @edolstra [pointed out](https://github.com/NixOS/nix/pull/8120#discussion_r1187413500) we have to take care of not occupying the `_type` namespace haphazardly, and also look out for round-tripping TOML correctly. We didn't come to a conclusion, and still owe a clear decision. Suggestions how to proceed are welcome.

<details><summary>Complete discussion</summary>

- @edolstra: the attrset type should say `rawTOML`
    - since `_type` is in a global namespace, no one else will be able to implement timestamps on top of that
    - @roberth: agreed. calling it timestamp comes with expectations we may have to maintain code around
        - the only objection may be that it's a scary name
        - @edolstra: all we have to express that it will be returned verbatim when serialising TOML again
- @edolstra: on the other hand, our TOML library can represent the different TOML time types, so maybe we can express this in the output
- @roberth: not having a `toTOML` is a bit of a liability
    - we wanted round-tripping, but currently have no way to test correctness
- needs more discussion about how to proceed

</details>

**Comment by andreabedini** [comment]
Thank you @fricklerhandwerk. I followed the rest of the code w.r.t. the experimental flag. Let me know if this is ok. I believe the CI failure is unrelated to the PR.

File: src/libutil/experimental-features.cc [hunk 2]
```diff
@@ -214,6 +214,13 @@ constexpr std::array<ExperimentalFeatureDetails, 13> xpFeatureDetails = {{
                 derivations that are themselves derivations outputs.
         )",
     },
+    {
+        .tag = Xp::ParseTomlTimestamps,
+        .name = "parse-toml-timestamps",
+        .description = R"(
+            Allow parsing of timestamps in builtins.fromTOML.
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply src/libutil/experimental-features.cc:221]: No problem :relaxed:

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting 2023-06-05:

- @fricklerhandwerk: shall we indeed make only the timestamp parsing experimental?
    - @ericson2314: can we deprecate the builtin altogether?
    - @edolstra: no does not make sense write parsers as Nix language libraries
- decision:
  - keep only the timestamp parsing experimental
  - merge

@andreabedini We'd very much welcome a follow up where timestamps are represented as structured Nix values. That would be the precondition to stabilise the feature.

**Pull request merged**